## Phần 1: Thay đổi cấu trúc ANN
### Yêu cầu
1. **Tăng số nút trong lớp ẩn**:  
   - Sửa lớp ẩn từ 4 nút thành 8 nút trong code định nghĩa `ANN`.  
   - Huấn luyện lại mô hình với cùng dữ liệu `X_train`, `y_train` từ lab (100 epochs).  
   - Ghi lại giá trị mất mát cuối cùng (`loss`) và độ chính xác trên tập kiểm tra (`X_test`, `y_test`).  

2. **Thêm một lớp ẩn**:  
   - Thêm một lớp ẩn thứ hai với 6 nút, dùng ReLU làm hàm kích hoạt. Cấu trúc mới:  
     - Đầu vào: 2 nút.  
     - Lớp ẩn 1: 8 nút (ReLU).  
     - Lớp ẩn 2: 6 nút (ReLU).  
     - Đầu ra: 1 nút (Sigmoid).  
   - Huấn luyện lại mô hình (100 epochs).  
   - Ghi lại giá trị mất mát cuối cùng và độ chính xác.

3. **Câu hỏi**:  
   - So sánh kết quả của 3 mô hình (4 nút, 8 nút, 8+6 nút):  
     - Mất mát cuối cùng thay đổi thế nào?  
     - Độ chính xác có cải thiện không? Tại sao bạn nghĩ vậy?  
   - Viết câu trả lời trong ô Markdown.

In [25]:
import numpy as np
import torch
import torch.nn as nn
import torch.optim as optim
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score
import matplotlib.pyplot as plt

# Giả lập dữ liệu ngẫu nhiên
np.random.seed(42)
X = np.random.rand(1000, 2)
y = (X[:, 0] + X[:, 1] > 1).astype(np.float32).reshape(-1, 1)

# Chia dữ liệu thành tập huấn luyện và tập kiểm tra
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Chuyển đổi sang tensor
X_train = torch.tensor(X_train, dtype=torch.float32)
y_train = torch.tensor(y_train, dtype=torch.float32)
X_test = torch.tensor(X_test, dtype=torch.float32)
y_test = torch.tensor(y_test, dtype=torch.float32)

# Định nghĩa mô hình ANN với 2 lớp ẩn
class ANN(nn.Module):
    def __init__(self):
        super(ANN, self).__init__()
        self.layer1 = nn.Linear(2, 8)  # Lớp ẩn 1 (8 nút)
        self.relu1 = nn.ReLU()
        self.layer2 = nn.Linear(8, 6)  # Lớp ẩn 2 (6 nút)
        self.relu2 = nn.ReLU()
        self.layer3 = nn.Linear(6, 1)  # Lớp đầu ra (1 nút)
        self.sigmoid = nn.Sigmoid()

    def forward(self, x):
        x = self.layer1(x)
        x = self.relu1(x)
        x = self.layer2(x)
        x = self.relu2(x)
        x = self.layer3(x)
        x = self.sigmoid(x)
        return x

# Khởi tạo mô hình
model = ANN()

# Định nghĩa hàm mất mát và bộ tối ưu hóa
criterion = nn.BCELoss()
optimizer = optim.Adam(model.parameters(), lr=0.01)

# Huấn luyện mô hình
epochs = 100
losses = []
for epoch in range(epochs):
    model.train()
    optimizer.zero_grad()
    outputs = model(X_train)
    loss = criterion(outputs, y_train)
    loss.backward()
    optimizer.step()
    losses.append(loss.item())

    if (epoch + 1) % 20 == 0:
        print(f"Epoch [{epoch+1}/{epochs}], Loss: {loss.item():.4f}")

# Đánh giá mô hình trên tập kiểm tra
model.eval()
with torch.no_grad():
    y_pred = model(X_test)
    y_pred_class = (y_pred >= 0.5).float()
    accuracy = accuracy_score(y_test, y_pred_class)
    print(f"Loss cuối cùng: {loss.item():.4f}")
    print(f"Độ chính xác trên tập kiểm tra: {accuracy:.4f}")


Epoch [20/100], Loss: 0.5520
Epoch [40/100], Loss: 0.3476
Epoch [60/100], Loss: 0.1810
Epoch [80/100], Loss: 0.1062
Epoch [100/100], Loss: 0.0752
Loss cuối cùng: 0.0752
Độ chính xác trên tập kiểm tra: 1.0000


In [26]:
import numpy as np
import torch
import torch.nn as nn
import torch.optim as optim
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score
import matplotlib.pyplot as plt

# Giả lập dữ liệu ngẫu nhiên
np.random.seed(42)
X = np.random.rand(1000, 2)
y = (X[:, 0] + X[:, 1] > 1).astype(np.float32).reshape(-1, 1)

# Chia dữ liệu thành tập huấn luyện và tập kiểm tra
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Chuyển đổi sang tensor
X_train = torch.tensor(X_train, dtype=torch.float32)
y_train = torch.tensor(y_train, dtype=torch.float32)
X_test = torch.tensor(X_test, dtype=torch.float32)
y_test = torch.tensor(y_test, dtype=torch.float32)

# Định nghĩa mô hình ANN với 2 lớp ẩn
class ANN(nn.Module):
    def __init__(self):
        super(ANN, self).__init__()
        self.layer1 = nn.Linear(2, 8)  # Lớp ẩn 1 (8 nút)
        self.relu1 = nn.ReLU()

        self.layer2 = nn.Linear(8, 1)  # Lớp đầu ra (1 nút)
        self.sigmoid = nn.Sigmoid()

    def forward(self, x):
        x = self.layer1(x)
        x = self.relu1(x)

        x = self.layer2(x)
        x = self.sigmoid(x)
        return x

# Khởi tạo mô hình
model = ANN()

# Định nghĩa hàm mất mát và bộ tối ưu hóa
criterion = nn.BCELoss()
optimizer = optim.Adam(model.parameters(), lr=0.01)

# Huấn luyện mô hình
epochs = 100
losses = []
for epoch in range(epochs):
    model.train()
    optimizer.zero_grad()
    outputs = model(X_train)
    loss = criterion(outputs, y_train)
    loss.backward()
    optimizer.step()
    losses.append(loss.item())

    if (epoch + 1) % 20 == 0:
        print(f"Epoch [{epoch+1}/{epochs}], Loss: {loss.item():.4f}")

# Đánh giá mô hình trên tập kiểm tra
model.eval()
with torch.no_grad():
    y_pred = model(X_test)
    y_pred_class = (y_pred >= 0.5).float()
    accuracy = accuracy_score(y_test, y_pred_class)
    print(f"Loss cuối cùng: {loss.item():.4f}")
    print(f"Độ chính xác trên tập kiểm tra: {accuracy:.4f}")


Epoch [20/100], Loss: 0.6081
Epoch [40/100], Loss: 0.5022
Epoch [60/100], Loss: 0.3883
Epoch [80/100], Loss: 0.2960
Epoch [100/100], Loss: 0.2329
Loss cuối cùng: 0.2329
Độ chính xác trên tập kiểm tra: 0.9700


In [27]:
import numpy as np
import torch
import torch.nn as nn
import torch.optim as optim
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score
import matplotlib.pyplot as plt

# Giả lập dữ liệu ngẫu nhiên
np.random.seed(42)
X = np.random.rand(1000, 2)
y = (X[:, 0] + X[:, 1] > 1).astype(np.float32).reshape(-1, 1)

# Chia dữ liệu thành tập huấn luyện và tập kiểm tra
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Chuyển đổi sang tensor
X_train = torch.tensor(X_train, dtype=torch.float32)
y_train = torch.tensor(y_train, dtype=torch.float32)
X_test = torch.tensor(X_test, dtype=torch.float32)
y_test = torch.tensor(y_test, dtype=torch.float32)

# Định nghĩa mô hình ANN với 2 lớp ẩn
class ANN(nn.Module):
    def __init__(self):
        super(ANN, self).__init__()
        self.layer1 = nn.Linear(2, 4)  # Lớp ẩn 1 (8 nút)
        self.relu1 = nn.ReLU()

        self.layer2 = nn.Linear(4, 1)  # Lớp đầu ra (1 nút)
        self.sigmoid = nn.Sigmoid()

    def forward(self, x):
        x = self.layer1(x)
        x = self.relu1(x)

        x = self.layer2(x)
        x = self.sigmoid(x)
        return x

# Khởi tạo mô hình
model = ANN()

# Định nghĩa hàm mất mát và bộ tối ưu hóa
criterion = nn.BCELoss()
optimizer = optim.Adam(model.parameters(), lr=0.01)

# Huấn luyện mô hình
epochs = 100
losses = []
for epoch in range(epochs):
    model.train()
    optimizer.zero_grad()
    outputs = model(X_train)
    loss = criterion(outputs, y_train)
    loss.backward()
    optimizer.step()
    losses.append(loss.item())

    if (epoch + 1) % 20 == 0:
        print(f"Epoch [{epoch+1}/{epochs}], Loss: {loss.item():.4f}")

# Đánh giá mô hình trên tập kiểm tra
model.eval()
with torch.no_grad():
    y_pred = model(X_test)
    y_pred_class = (y_pred >= 0.5).float()
    accuracy = accuracy_score(y_test, y_pred_class)
    print(f"Loss cuối cùng: {loss.item():.4f}")
    print(f"Độ chính xác trên tập kiểm tra: {accuracy:.4f}")


Epoch [20/100], Loss: 0.6598
Epoch [40/100], Loss: 0.6063
Epoch [60/100], Loss: 0.5334
Epoch [80/100], Loss: 0.4584
Epoch [100/100], Loss: 0.3914
Loss cuối cùng: 0.3914
Độ chính xác trên tập kiểm tra: 0.8750


###Kết luận: Cả 3 trường hợp(4 nút, 8 nút, 8+6 nút đều có sự chênh lệch mất mát một cách đáng kể.

- Chạy 3 lần cho thấy độ chính xác cuối cùng về độ cải thiện lúc có lúc không, bởi vì giá trị ở mỗi con số đầu vào và đầu ra càng nhiều độ chính xác có lúc lên thì sẽ có lúc xuống


## Phần 2: Thử nghiệm với hàm mất mát và tối ưu hóa
### Yêu cầu
1. **Dùng BCEWithLogitsLoss thay cho BCELoss**:  
   - Thay `nn.BCELoss()` bằng `nn.BCEWithLogitsLoss()`.  
   - Xóa hàm Sigmoid khỏi lớp đầu ra của mô hình (vì `BCEWithLogitsLoss` tự xử lý).  
   - Huấn luyện lại mô hình với cấu trúc ban đầu (2-4-1, 100 epochs).  
   - Ghi lại mất mát cuối cùng và độ chính xác.

2. **Thay Adam bằng SGD**:  
   - Dùng lại cấu trúc ban đầu (2-4-1) với `nn.BCELoss()`.  
   - Thay `optim.Adam` bằng `optim.SGD` với `lr=0.01`.  
   - Huấn luyện lại (100 epochs).  
   - Ghi lại mất mát cuối cùng và độ chính xác.

3. **Câu hỏi**:  
   - So sánh kết quả:  
     - `BCEWithLogitsLoss` có khác gì so với `BCELoss` về mất mát và độ chính xác? Tại sao?  
     - `SGD` so với `Adam`: Mất mát giảm nhanh hơn hay chậm hơn? Độ chính xác thay đổi ra sao?  
   - Viết câu trả lời trong ô Markdown.

In [29]:
import numpy as np
import torch
import torch.nn as nn
import torch.optim as optim
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score
import matplotlib.pyplot as plt

# Giả lập dữ liệu ngẫu nhiên
np.random.seed(42)
X = np.random.rand(1000, 2)
y = (X[:, 0] + X[:, 1] > 1).astype(np.float32).reshape(-1, 1)

# Chia dữ liệu thành tập huấn luyện và tập kiểm tra
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Chuyển đổi sang tensor
X_train = torch.tensor(X_train, dtype=torch.float32)
y_train = torch.tensor(y_train, dtype=torch.float32)
X_test = torch.tensor(X_test, dtype=torch.float32)
y_test = torch.tensor(y_test, dtype=torch.float32)

# Định nghĩa mô hình ANN với 2 lớp ẩn
class ANN(nn.Module):
    def __init__(self):
        super(ANN, self).__init__()
        self.layer1 = nn.Linear(2, 4)  # Lớp ẩn 1 (8 nút)
        self.relu1 = nn.ReLU()

        self.layer2 = nn.Linear(4, 1)  # Lớp đầu ra (1 nút)

    def forward(self, x):
        x = self.layer1(x)
        x = self.relu1(x)

        x = self.layer2(x)

        return x

# Khởi tạo mô hình
model = ANN()

# Định nghĩa hàm mất mát và bộ tối ưu hóa
criterion = nn.BCEWithLogitsLoss()
optimizer = optim.Adam(model.parameters(), lr=0.01)

# Huấn luyện mô hình
epochs = 100
losses = []
for epoch in range(epochs):
    model.train()
    optimizer.zero_grad()
    outputs = model(X_train)
    loss = criterion(outputs, y_train)
    loss.backward()
    optimizer.step()
    losses.append(loss.item())

    if (epoch + 1) % 20 == 0:
        print(f"Epoch [{epoch+1}/{epochs}], Loss: {loss.item():.4f}")

# Đánh giá mô hình trên tập kiểm tra
model.eval()
with torch.no_grad():
    y_pred = model(X_test)
    y_pred_class = (y_pred >= 0.5).float()
    accuracy = accuracy_score(y_test, y_pred_class)
    print(f"Loss cuối cùng: {loss.item():.4f}")
    print(f"Độ chính xác trên tập kiểm tra: {accuracy:.4f}")


Epoch [20/100], Loss: 0.6343
Epoch [40/100], Loss: 0.5769
Epoch [60/100], Loss: 0.4954
Epoch [80/100], Loss: 0.4106
Epoch [100/100], Loss: 0.3383
Loss cuối cùng: 0.3383
Độ chính xác trên tập kiểm tra: 0.9150


###Mất mát cuối cùng: 0.3383
  Độ chính xác: 0.9150

In [32]:
import numpy as np
import torch
import torch.nn as nn
import torch.optim as optim
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score
import matplotlib.pyplot as plt

# Giả lập dữ liệu ngẫu nhiên
np.random.seed(42)
X = np.random.rand(1000, 2)
y = (X[:, 0] + X[:, 1] > 1).astype(np.float32).reshape(-1, 1)

# Chia dữ liệu thành tập huấn luyện và tập kiểm tra
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Chuyển đổi sang tensor
X_train = torch.tensor(X_train, dtype=torch.float32)
y_train = torch.tensor(y_train, dtype=torch.float32)
X_test = torch.tensor(X_test, dtype=torch.float32)
y_test = torch.tensor(y_test, dtype=torch.float32)

# Định nghĩa mô hình ANN với 2 lớp ẩn
class ANN(nn.Module):
    def __init__(self):
        super(ANN, self).__init__()
        self.layer1 = nn.Linear(2, 4)  # Lớp ẩn 1 (8 nút)
        self.relu1 = nn.ReLU()

        self.layer2 = nn.Linear(4, 1)  # Lớp đầu ra (1 nút)
        self.sigmoid = nn.Sigmoid()
    def forward(self, x):
        x = self.layer1(x)
        x = self.relu1(x)

        x = self.layer2(x)
        x= self.sigmoid(x)
        return x

# Khởi tạo mô hình
model = ANN()

# Định nghĩa hàm mất mát và bộ tối ưu hóa
criterion = nn.BCELoss()
optimizer = optim.SGD(model.parameters(), lr=0.01)

# Huấn luyện mô hình
epochs = 100
losses = []
for epoch in range(epochs):
    model.train()
    optimizer.zero_grad()
    outputs = model(X_train)
    loss = criterion(outputs, y_train)
    loss.backward()
    optimizer.step()
    losses.append(loss.item())

    if (epoch + 1) % 20 == 0:
        print(f"Epoch [{epoch+1}/{epochs}], Loss: {loss.item():.4f}")

# Đánh giá mô hình trên tập kiểm tra
model.eval()
with torch.no_grad():
    y_pred = model(X_test)
    y_pred_class = (y_pred >= 0.5).float()
    accuracy = accuracy_score(y_test, y_pred_class)
    print(f"Loss cuối cùng: {loss.item():.4f}")
    print(f"Độ chính xác trên tập kiểm tra: {accuracy:.4f}")

Epoch [20/100], Loss: 0.7131
Epoch [40/100], Loss: 0.7060
Epoch [60/100], Loss: 0.7001
Epoch [80/100], Loss: 0.6951
Epoch [100/100], Loss: 0.6910
Loss cuối cùng: 0.6910
Độ chính xác trên tập kiểm tra: 0.5300


###Loss cuối cùng: 0.6910
Độ chính xác trên tập kiểm tra: 0.5300

###So sánh kết quả:
BCEWithLogitsLoss có khác gì so với BCELoss về mất mát và độ chính xác? Tại sao?
SGD so với Adam: Mất mát giảm nhanh hơn hay chậm hơn? Độ chính xác thay đổi ra sao?
Viết câu trả lời trong ô Markdown.

Về mất mát giữa BCEWithLogitsLoss và BCELoss thì BCELoss chạy mất mát cao hơn, mất mát giảm nhanh hơn, độ chính xác giảm đi đáng kể